contracts/upgradeable_contracts/multi_amb_erc20_to_erc677/MultiTokenForwardingRules.sol

```diff
@@ -0,0 +1,76 @@
+pragma solidity 0.4.24;
+
+import "../../upgradeable_contracts/Ownable.sol";
+
+/**
+ * @title MultiTokenForwardingRules
+ * @dev Multi token mediator functionality for managing destination AMB lanes permissions.
+ */
+contract MultiTokenForwardingRules is Ownable {
+    address public constant ANY_ADDRESS = 0xFFfFfFffFFfffFFfFFfFFFFFffFFFffffFfFFFfF;
+
+    event ForwardingRuleUpdated(address token, address sender, address receiver, int256 lane);
+
+    /**
+     * @dev Checks the destination lane for a particular bridge operation.
+     * Makes a top-down checking of forwarding rules, from specific ones to more generic ones.
+     * @param _token address of the token contract on the foreign side of the bridge.
+     * @param _sender address of the tokens sender on the home side of the bridge.
+     * @param _receiver address of the tokens receiver on the foreign side of the bridge.
+     * @return destination lane identifier, where the message should be forwarded to.
+     *  1 - oracle-driven-lane should be used.
+     *  0 - default behaviour should be applied.
+     * -1 - manual lane should be used.
+     */
+    function destinationLane(address _token, address _sender, address _receiver) public view returns (int256) {
```

Comment:
This approach is very straightforward, thanks! But the main drawback here is excessive gas usage: for the worst case it will require 8 RSTORE and 8 keccak operations.

Can we optimize it a bit?

1. `forwardingRule(ANY_ADDRESS, ANY_ADDRESS, ANY_ADDRESS)` --  will be always 0 for the multi-token extension between Ethereum and xDai: we either subsidize its operations or not -- it will be configured through the AL on the oracle side

2. `forwardingRule(_token, _sender, _receiver)` and `forwardingRule(ANY_ADDRESS, _sender, _receiver)` are too specific and I cannot see any useful case when we could enable it.

3. As per the original requirements a positive returned value of `forwardingRule(_token, _sender, ANY_ADDRESS)` or `forwardingRule(_token, ANY_ADDRESS, _receiver)` makes sense only if `forwardingRule(_token, ANY_ADDRESS, ANY_ADDRESS)` returns a negative value. The case when `forwardingRule(_token, ANY_ADDRESS, ANY_ADDRESS)` returns a positive value is not going to be used.

Reply:
Agree on that, I optimized the number of used of storage read operations according to your comments.
Please, take a look.

Reply:
Do you think that this snippet will be closer to the initial requirements?

```
lane = forwardingRule(_token, ANY_ADDRESS, ANY_ADDRESS);
if (lane < 0) {
    int256 lane_n = forwardingRule(_token, _sender, ANY_ADDRESS);
    if (lane_n > 0) return 1;
    lane_n = forwardingRule(_token, ANY_ADDRESS, _receiver);
    if (lane_n > 0) return 1;
}
if (lane != 0) return lane;
```

In this case it seems logical to introduce `setTokenFwdRule`, `setSenderExceptionForTokenFwdRule`, `setReceiverExceptionForTokenFwdRule`, `setSenderFwdRule` and `setReceiverFwdRule`. They will call `setForwardingRule` (it should be internal in this case) and force: 
  * `setForwardingRule` to be called with `-1` in `setTokenFwdRule`
  * `setForwardingRule` to be called with `1` in `setSenderExceptionForTokenFwdRule`
  * `setForwardingRule` to be called with `1` in `setReceiverExceptionForTokenFwdRule`
  * `setForwardingRule` to be called with `-1` in ``setSenderFwdRule`
  * `setForwardingRule` to be called with `-1` in ``setReceiverFwdRule`

What do you think?

Reply:
How does this snippet handles the case with inclusion rules?

Reply:
Sorry, I missed the commit "Reduce amount of SLOAD operations required" where you already implemented this logic. It looks very similar to what I am suggesting:
```
lane = forwardingRule(_token, ANY_ADDRESS, ANY_ADDRESS);
if (lane < 0) {
    int256 lane_n = forwardingRule(_token, _sender, ANY_ADDRESS);
    if (lane_n > 0) return 1;
    lane_n = forwardingRule(_token, ANY_ADDRESS, _receiver);
    if (lane_n > 0) return 1;
}
if (lane != 0) return lane;

lane = forwardingRule(ANY_ADDRESS, _sender, ANY_ADDRESS); 
if (lane != 0) return lane;
return forwardingRule(ANY_ADDRESS, ANY_ADDRESS, _receiver);
```

Together with the logic I suggested for the methods `setTokenFwdRule`, `setSenderExceptionForTokenFwdRule`, `setReceiverExceptionForTokenFwdRule`, `setSenderFwdRule` and `setReceiverFwdRule` it could correspond to the requirements introduced initially.